conn,device: support conn.ReceiveFunc nonzero packet offset

jwhited opened the pull request:
This enables a conn.Bind to leverage wireguard-go packet memory for reading packets containing layers that wireguard-go should ignore, e.g. a VXLAN or Geneve header preceeding WireGuard.

sfllaw commented on conn/conn.go:
```diff
@@ -21,11 +21,13 @@ const (
-// and eps slice with a length greater than or equal to the length of packets.
-// These lengths must not exceed the length of the associated Bind.BatchSize().
-type ReceiveFunc func(packets [][]byte, sizes []int, eps []Endpoint) (n int, err error)
+// sizes, packets, and endpoints that should be evaluated. A sizes element
+// includes both the starting and ending offset for an element of packets. Some
```
:question: Confused: Is the `sizes` element defined as `[start, end)`? It would be nice if the documentation were more explicit about ranges.

:question: Confused: Is this the name `sizes` still the right name for this, now that it is some kind of range or interval? Granted, I cannot come up with a better name.

sfllaw commented on conn/bind_std.go:
```diff
@@ -231,7 +231,7 @@ func (s *StdNetBind) receiveIP(
 	rxOffload bool,
 	bufs [][]byte,
-	sizes []int,
+	sizes [][2]int,
```
:question: Confused: Is this tuple always going to be a range? Would it be more clear as a `type size struct{ start, end int }` or is that worse somehow?

jwhited replied:
Practically speaking I believe the size and layout would be the same. I agree that a struct is more readable. We are also balancing with limiting change (literal LoC) to reduce drift from upstream before larger refactors are appropriate.

sfllaw reviewed (commented):
Just some notes about sources of confusion from someone unfamiliar with this codebase. Feel free to ignore.

danderson commented on conn/bind_std.go:
```diff
@@ -268,8 +268,8 @@ func (s *StdNetBind) receiveIP(
 		msg := &(*msgs)[i]
-		sizes[i] = msg.N
-		if sizes[i] == 0 {
+		sizes[i][0], sizes[i][1] = 0, msg.N
```
minor suggestion, optional: move the continue branch below further up, and make the condition look at msg.N instead of the sizes array? As written here it took me a beat to translate the branch condition to "oh, skip empty messages".

danderson reviewed (approved):
LGTM. Reflecting some discussion off-github: we're balancing extending the functionality of this API for a pressing need, vs. doing a larger scale refactor to improve the packet wrangling API as a whole.

The latter is tempting, but is likely to spiral into a large change as it bubbles through more of the API. We'd rather tackle a broader redesign separately (and we're gearing up to, available bandwidth notwithstanding), rather than end up pulling refactors into this targeted feature change.

jwhited committed:
conn,device: support conn.ReceiveFunc nonzero packet offset

This enables a conn.Bind to leverage wireguard-go packet memory for
reading packets containing layers that wireguard-go should ignore, e.g.
a VXLAN or Geneve header preceeding WireGuard.

Signed-off-by: Jordan Whited <jordan@tailscale.com>

jwhited commented:
As-is this commit doesn't account for the extra 8 bytes on the TUN write side of the packet pipeline, which is a bug. Tests pass because there is no test exercising a nonzero receive offset.

Addressing this likely requires adding a field to `QueueInboundElement` accounting for the offset, as a `conn.Bind` may selectively pass a nonzero offset. We don't want to always assume an extra 8 bytes at the packet head. A `conn.Bind` may receive packets with or without an encapsulating transport on the same socket/ReceiveFunc.

Closing this for now until there's time to revisit. We can have the `conn.Bind` copy/memmove away the Geneve header for now until that performance penalty is worth addressing.
